loader/unknown_function_handling.c

```diff
@@ -0,0 +1,322 @@
+                if (strcmp(property->entrypoints[entry], funcName) == 0) {
+                    return true;
+                }
+            }
+        }
+    }
+
+    return false;
+}
+
+void loader_free_dev_ext_table(struct loader_instance *inst) {
+    for (uint32_t i = 0; i < inst->dev_ext_disp_function_count; i++) {
+        loader_instance_heap_free(inst, inst->dev_ext_disp_functions[i]);
+    }
+    memset(inst->dev_ext_disp_functions, 0, sizeof(inst->dev_ext_disp_functions));
+}
+
+/*
+ * This function returns generic trampoline code address for unknown entry points.
+ * Presumably, these unknown entry points (as given by funcName) are device
+ * extension entrypoints.  An array is used to keep a list of unknown entry
+ * points and their mapping to the device extension dispatch table
+ * (struct loader_dev_ext_dispatch_table).
+ * \returns
+ * For a given entry point string (funcName), if an existing mapping is found
+ * the  * trampoline address for that mapping is returned. Otherwise, this unknown
+ * entry point has not been seen yet. Next check if a layer or ICD supports it.  If so then
+ * a  * new entry in the hash table is initialized and that trampoline address for
+ * the new entry is returned. Null is returned if the hash table is full or
+ * if no discovered layer or ICD returns a non-NULL GetProcAddr for it.
+ */
+void *loader_dev_ext_gpa(struct loader_instance *inst, const char *funcName) {
+    // Linearly look through already added functions to make sure we haven't seen it before
+    // if we have, return the function at the index found
+    for (uint32_t i = 0; i < inst->dev_ext_disp_function_count; i++) {
+        if (inst->dev_ext_disp_functions[i] && !strcmp(inst->dev_ext_disp_functions[i], funcName))
+            return loader_get_dev_ext_trampoline(i);
+    }
+
+    // Check if funcName is supported in either ICDs or a layer library
+    if (!loader_check_icds_for_dev_ext_address(inst, funcName) &&
+        !loader_check_layer_list_for_dev_ext_address(&inst->app_activated_layer_list, funcName)) {
+        // if support found in layers continue on
+        return NULL;
+    }
+
+    // add found function to dev_ext_disp_functions;
+    size_t funcName_len = strlen(funcName) + 1;
+    inst->dev_ext_disp_functions[inst->dev_ext_disp_function_count] =
+        (char *)loader_instance_heap_alloc(inst, funcName_len, VK_SYSTEM_ALLOCATION_SCOPE_INSTANCE);
+    if (NULL == inst->dev_ext_disp_functions[inst->dev_ext_disp_function_count]) {
+        // failed to allocate memory, return NULL
+        return NULL;
+    }
+    strncpy(inst->dev_ext_disp_functions[inst->dev_ext_disp_function_count], funcName, funcName_len);
+    // init any dev dispatch table entries as needed
+    loader_init_dispatch_dev_ext_entry(inst, NULL, inst->dev_ext_disp_function_count, funcName);
+    void *out_function = loader_get_dev_ext_trampoline(inst->dev_ext_disp_function_count);
+    inst->dev_ext_disp_function_count++;
```

Comment:
Should we add an assert in all these cases for the scenario of adding too many functions?

Reply:
Very good catch. Now it logs an ERROR and returns NULL.